crates/matrix-sdk-crypto/src/gossiping/machine.rs

```diff
@@ -315,76 +315,74 @@ impl GossipMachine {
 
-        if let Some(device) = device {
-            match self.should_share_key(&device, &session).await {
-                Err(e) => {
-                    if let KeyForwardDecision::ChangedSenderKey = e {
-                        warn!(
-                            user_id = device.user_id().as_str(),
-                            device_id = device.device_id().as_str(),
-                            "Received a key request from a device that changed \
-                            their Curve25519 sender key"
-                        );
-                    } else {
-                        debug!(
-                            user_id = device.user_id().as_str(),
-                            device_id = device.device_id().as_str(),
-                            reason = ?e,
-                            "Received a key request that we won't serve",
-                        );
-                    }
-
-                    Ok(None)
-                }
-                Ok(message_index) => {
-                    info!(
-                        user_id = %device.user_id(),
-                        device_id = %device.device_id(),
-                        session_id = session.session_id(),
-                        room_id = %session.room_id,
-                        ?message_index,
-                        "Serving a room key request",
-                    );
-
-                    match self.forward_room_key(&session, &device, message_index).await {
-                        Ok(s) => Ok(Some(s)),
-                        Err(OlmError::MissingSession) => {
-                            info!(
-                                user_id = %device.user_id(),
-                                device_id = %device.device_id(),
-                                session_id = session.session_id(),
-                                "Key request is missing an Olm session, \
-                                putting the request in the wait queue",
-                            );
-                            self.handle_key_share_without_session(device, event.to_owned().into());
-
-                            Ok(None)
-                        }
-                        Err(OlmError::SessionExport(e)) => {
-                            warn!(
-                                user_id = %device.user_id(),
-                                device_id = %device.device_id(),
-                                session_id = session.session_id(),
-                                "Can't serve a room key request, the session \
-                                can't be exported into a forwarded room key: \
-                                {:?}",
-                                e
-                            );
-                            Ok(None)
-                        }
-                        Err(e) => Err(e),
-                    }
-                }
-            }
-        } else {
+        let Some(device) = device else {
```

Comment:
I think I can get behind the device `None` case early return, but for the rest I would prefer the whole `match self.forward_room_key(&session, &device, message_index).await` block to become a separate method.

Alternatively, just leaving out the second early return works for me as well.

Reply:
I don't understand exactly what you mean. Can you elaborate, or just push a commit with your preferred solution?

Reply:
Did so in https://github.com/matrix-org/matrix-rust-sdk/pull/1185/commits/852d16d09224e040d1dcc4f98f28d9ded5ae0522.